WcaOnRails/app/javascript/edit-schedule/SchedulesEditor/calendar-utils.js

```
@@ -0,0 +1,230 @@
+      newActivity.endTime = momentToIso(newEnd);
+    } else {
+      // Do nothing, user cliked an event without any event selected.
+      return;
+    }
+    roomSelected.activities.push(newActivity);
+    if (renderItOnCalendar) {
+      // Cloning the object here: activityToFcEvent modifies in place,
+      // and we don't want event.selected to propagate in the WCIF!
+      let fcEvent = Object.assign({}, newActivity);
+      fcEvent = activityToFcEvent(fcEvent);
+      singleSelectEvent(fcEvent);
+      $(scheduleElementSelector).fullCalendar("renderEvent", fcEvent);
+    }
+    // update list of activityCode used, and rootRender to display the save message
+    reactElem.setState({ usedActivityCodeList: [...reactElem.state.usedActivityCodeList, newActivity.activityCode] }, rootRender());
+  }
+}
+
+function handleEventModifiedInCalendar(reactElem, event) {
+  let room = roomWcifFromId(reactElem.props.scheduleWcif, reactElem.state.selectedRoom);
+  let activityIndex = activityIndexInArray(room.activities, event.id);
+  if (activityIndex < 0) {
+    alert("This is very very BAD, I couldn't find an activity matching the modified event!");
+  }
+  let activity = room.activities[activityIndex];
+  activity.name = event.name;
+  activity.activityCode = event.activityCode;
+  activity.startTime = momentToIso(event.start);
+  activity.endTime = momentToIso(event.end);
+  // We rootRender to display the "Please save your changes..." message
+  rootRender();
+}
+
+function handleRemoveEventFromCalendar(reactElem, event) {
+  if (!confirm(`Are you sure you want to remove ${event.name}`)) {
+    return false;
+  }
+
+  // Remove activityCode from the list used by the ActivityPicker
+  let newActivityCodeList = reactElem.state.usedActivityCodeList;
+  let activityCodeIndex = newActivityCodeList.indexOf(event.activityCode);
+  if (activityCodeIndex < 0) {
+    alert("This is BAD, I couldn't find an activity code when removing event!");
```

As I mentioned before, it's probably better to throw an error here instead.